cylc/flow/scheduler.py

```diff
@@ -747,12 +748,37 @@ async def run(self):
         """Run the startup sequence and set the main loop running.
 
         Lightweight wrapper for testing convenience.
-
         """
+        # register signal handlers
+        for sig in (
+            # ctrl+c
+            signal.SIGINT,
+            # the "stop please" signal
+            signal.SIGTERM,
+            # the controlling process bailed
+            # (e.g. terminal quit in no detach mode)
+            signal.SIGHUP,
+        ):
+            signal.signal(sig, self._handle_signal)
+
         await self.start()
         # note run_scheduler handles its own shutdown logic
         await self.run_scheduler()
 
+    def _handle_signal(self, sig: int, frame) -> None:
```

Comment:
Possible test

```python
@pytest.mark.parametrize("signal", ((SIGHUP), (SIGINT), (SIGTERM)))
async def test_eg(one, start, signal, log_filter):
    async with start(one) as log:
        one._handle_signal(signal, None)
        assert log_filter(log, contains=signal.name)
        assert one.stop_mode.name == 'REQUEST_NOW'
        one._handle_signal(signal, None)
        assert one.stop_mode.name == 'REQUEST_NOW_NOW'
```

Reply:
The functional tests cover most of this, but don't cover the NOW/NOW_NOW escalation (as you've spotted).

Will do.